src/storage/checkpoint_io.cpp

```diff
@@ -0,0 +1,60 @@
+#include "storage/checkpoint_io.h"
+#include <vector>
+namespace terrier::storage {
+
+void BufferedTupleWriter::SerializeTuple(ProjectedColumns::RowView *row, ProjectedRow *row_buffer,
+                                         const storage::BlockLayout &layout) {
+  // First construct tuple in row_buffer, and calculate total size required
+  int32_t varlen_offset = 0;
+  std::vector<const VarlenEntry *> varlen_entries;
+  for (uint16_t projection_list_idx = 0; projection_list_idx < row->NumColumns(); projection_list_idx++) {
+    if (row->IsNull(projection_list_idx)) {
+      row_buffer->SetNull(projection_list_idx);
+    } else {
+      row_buffer->SetNotNull(projection_list_idx);
+      storage::col_id_t col = row->ColumnIds()[projection_list_idx];
+      if (layout.IsVarlen(col)) {
+        const VarlenEntry *varlen_entry = reinterpret_cast<VarlenEntry *>(row->AccessForceNotNull(projection_list_idx));
+        if (varlen_entry->IsInlined()) {
+          std::memcpy(row_buffer->AccessForceNotNull(projection_list_idx), varlen_entry, sizeof(VarlenEntry));
+        } else {
+          // use the content_ field in VarlenEntry as the offset to the varlen file.
+          uint32_t size = varlen_entry->Size();
+          *reinterpret_cast<VarlenEntry *>(row_buffer->AccessForceNotNull(projection_list_idx)) =
+              VarlenEntry::CreateCheckpoint(varlen_offset, size);
+          varlen_entries.push_back(varlen_entry);
+          varlen_offset += (static_cast<uint32_t>(sizeof(uint32_t)) + varlen_entry->Size());
+        }
+      } else {
+        std::memcpy(row_buffer->AccessForceNotNull(projection_list_idx), row->AccessForceNotNull(projection_list_idx),
+                    layout.AttrSize(col));
+      }
+    }
+  }
+  AppendTupleToBuffer(row_buffer, varlen_offset, varlen_entries);
+}
+
+void BufferedTupleWriter::AppendTupleToBuffer(ProjectedRow *row_buffer, int32_t total_varlen,
+                                              const std::vector<const VarlenEntry *> &varlen_entries) {
+  int32_t row_size = row_buffer->Size();
+  int32_t tot_size = total_varlen + row_size;
+  // Flush buffer first if the tuple cannot fit in buffer
+  if (cur_buffer_size_ + tot_size > block_size_) {
+    PersistBuffer();
```

Comment:
This is blocking, can't you get a new buffer and add to that while this is asynchronously persisting in the background

Reply:
Same as https://github.com/cmu-db/terrier/pull/341/files#r274158232